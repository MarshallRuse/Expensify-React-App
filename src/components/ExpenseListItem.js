import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = (props) => (
    <div>
        <Link to={`/edit/${props.id}`}><h2>{props.description}</h2></Link>
        <h3>
            Amount: {numeral(props.amount / 100).format('$0,0.00')}
        </h3>
        <p>
            Created: {moment(props.createdAt).format('MMMM Do, YYYY')}
        </p>
    </div>
);

export default ExpenseListItem;

