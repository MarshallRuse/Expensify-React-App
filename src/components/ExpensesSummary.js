import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../redux/selectors/expenses';
import selectExpensesTotal from '../redux/selectors/expenses-total';

export const ExpensesSummary = ({expensesCount, expensesTotal}) => {
    const expenseWord = expensesCount === 1 ? 'expense' : 'expenses';

    return (
    <div>
        <h2>Expenses Summary</h2>
        <h3>You have {expensesCount} {expenseWord} totaling {expensesTotal}</h3>
    </div>
)}


const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);

    return {
        expensesCount: visibleExpenses.length,
        expensesTotal: selectExpensesTotal(visibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary)

