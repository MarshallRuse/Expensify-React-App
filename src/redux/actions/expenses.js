 import uuid from 'uuid';
 
 // ADD_EXPENSE
export const addExpense = ({
    id = undefined,
    description = '',
    notes = '',
    amount = 0, 
    createdAt = 0
} = {}) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: id ? id : uuid(), // note this works here, not in the arguments
        description,
        notes,
        amount,
        createdAt
    }
});

    // REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
})

    // EDIT EXPENSE
export const editExpense = (id = '', updates = {}) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})