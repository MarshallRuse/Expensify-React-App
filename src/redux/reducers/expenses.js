const expensesReducerDefaultState = [];

const expensesReducer = ( state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            // .concat() doesn't mutate state directly like .push() would,
            // returns a new array
            return state.concat(action.expense);
        case 'REMOVE_EXPENSE':
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return { ...expense, ...action.updates};
                } else {
                    return expense;
                }
            })
        case 'SET_EXPENSES':
            return action.expenses;
        default:
            return state;
    }
}

export default expensesReducer;