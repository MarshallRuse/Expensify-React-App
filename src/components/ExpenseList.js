import React from 'react';
import {connect} from 'react-redux';

import selectExpenses from '../redux/selectors/expenses';

import ExpenseListItem from './ExpenseListItem';
import ExpenseListFilter from './ExpenseListFilter';


export const ExpenseList = (props) => (
    <div>
        {props.expenses.length > 0 && <ExpenseListFilter />}
        <div className='content-container'>
            <div className='list__header'>
                <div className='show-for-mobile'>Expenses</div>
                <div className='show-for-desktop'>Expense</div>
                <div className='show-for-desktop'>Amount</div>
            </div>
            { props.expenses.length === 0 ? (
                <div className='list__item--empty'>
                    <span>No Expenses</span>
                </div>
            ) : (
                
                props.expenses.map((expense) => <ExpenseListItem key={expense.id} {...expense}/>)
            ) 
            }
        </div>
    </div>
    
);

/*

*/

// Note, unlike ./playground/higher-order-components,
// connect() returns a function to be called with ExpenseList.
// Props can be passed as a callback function to connect as an argument

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
};

// By convention just default export the connect function
export default connect(mapStateToProps)(ExpenseList);
