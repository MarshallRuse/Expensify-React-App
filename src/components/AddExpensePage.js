import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../redux/actions/expenses';

export class AddExpensePage extends React.Component {

    onSubmit = (expense) => {
        // ** The below has been replaced due to mapDispatchToProps,
        // which allows for easier testing **

        // props.dispatch(addExpense(expense));
        this.props.startAddExpense(expense);

        // history.push is used to redirect programatically.
        // this comes from components used within react-router
        this.props.history.push('/dashboard');
    }

    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm 
                    onSubmit={this.onSubmit}    
                />
            </div>
        )
    }
}


// The second argument to connect(), mapDispatchToProps, 
// can be explicitly set so that onSubmit can be accessed by the 
// testing functions
const mapDispatchToProps = (dispatch) => {
    return {
        startAddExpense: (expense) => dispatch(startAddExpense(expense))
    }
}

export default connect(undefined, mapDispatchToProps)(AddExpensePage);