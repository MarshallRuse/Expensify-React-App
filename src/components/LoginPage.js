import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../redux/actions/auth';

export const LoginPage = (props) => (
    <div className='box-layout'>
        <div className='box-layout__box'>
            <h1 className='box-layout__title'>Expensify</h1>
            <p>Buy things, track them, realize how much you've been eating out...</p>
            <button className='button' onClick={props.startLogin}>Login with Google</button>
        </div>
    </div>
)

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
})

export default connect(undefined, mapDispatchToProps)(LoginPage);