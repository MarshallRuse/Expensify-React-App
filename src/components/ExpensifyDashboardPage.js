import React from 'react';

import ExpenseList from './ExpenseList';
import ExpensesSummary from './ExpensesSummary';

const ExpensifyDashboardPage = () => (
    <div>
        <ExpensesSummary />
        <ExpenseList />
    </div>
);

export default ExpensifyDashboardPage;