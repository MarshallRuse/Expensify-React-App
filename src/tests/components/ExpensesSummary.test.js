import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

let wrapper;

beforeEach(() => {
    wrapper = shallow(<ExpensesSummary expenses={[]}/>);
});


test('Should render the ExpensesSummary component with no expenses', () => {
    expect(wrapper).toMatchSnapshot();
});

test('Should render the ExpensesSummary component with expenses', () => {
    wrapper.setProps({ expenses });
    expect(wrapper).toMatchSnapshot();
});
