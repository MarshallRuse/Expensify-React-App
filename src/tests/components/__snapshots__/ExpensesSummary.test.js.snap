// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the ExpensesSummary component with expenses 1`] = `
<div>
  <h2>
    Expenses Summary
  </h2>
  <h3>
    Your total expenses amount is: 
    $1,141.95
  </h3>
</div>
`;

exports[`Should render the ExpensesSummary component with no expenses 1`] = `
<div>
  <h2>
    Expenses Summary
  </h2>
  <h3>
    Your total expenses amount is: 
    $0.00
  </h3>
</div>
`;
